Support CosmosDB and Redis role assignments on app specific managed identity

eerhardt opened the pull request:
## Description

We currently don't support an easy API for modifying the role assignments on database resources. There are no role types that can be used from the Azure.Provisioning libraries today. So for now we will just move our existing role assignments to the ACA roles bicep module.

To support database resources, we add a new virtual method on AzureProvisioningResource - `AddRoleAssignments` passing in a context object containing the information on the role assignments to add.

Also needed to split all role assignments out of the {app}-roles bicep module because of an issue with CosmosDB failing if the role assignment is added in the same bicep file as the managed identity it is pointing to. This simplifies the code because the role assignment generation is the same whether the resource is existing or not.

Contributes to #6161
Fix #8142

## Checklist

- Is this feature complete?
  - [x] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
- Did you add public API?
  - [x] Yes
    - If yes, did you have an API Review for it?
      - [x] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [x] Yes
- Does the change make any security assumptions or guarantees?
  - [x] Yes
    - If yes, have you done a threat model and had a security review?
      - Ongoing
- Does the change require an update in our Aspire docs?
  - [x] Yes
        - Link to aspire-docs issue [[New article]: Aspire 9.2 Azure WithRoleAssignments docs (dotnet/docs-aspire#2788)](https://github.com/dotnet/docs-aspire/issues/2788)

eerhardt committed:
Support CosmosDB and Redis role assignments on app specific managed identity

We currently don't support an easy API for modifying the role assignments on database resources. There are no role types that can be used from the Azure.Provisioning libraries today. So for now we will just move our existing role assignments to the ACA roles bicep module.

To support database resources, we add a new annotation "RoleAssignmentCustomizationAnnotation" which contains a callback that the resource can implement to inject the role assignment objects into the Azure bicep infrastructure.

Contributes to #6161

eerhardt committed:
Update RedisCacheAccessPolicyAssignment to set a unique name

Fix #8142

Also fix tests for new behavior

davidfowl commented on playground/Redis/Redis.ApiService/Program.cs:
```diff
@@ -1,14 +1,31 @@
-builder.AddKeyedRedisClient("garnet");
-builder.AddKeyedRedisClient("valkey");
+
+var azureOptionsProvider = new AzureOptionsProvider();
+
+var configurationOptions = ConfigurationOptions.Parse(
+    builder.Configuration.GetConnectionString("redis") ??
+    throw new InvalidOperationException("Could not find a 'redis' connection string."));
+
+if (configurationOptions.EndPoints.Any(azureOptionsProvider.IsMatch))
+{
+    await configurationOptions.ConfigureForAzureWithTokenCredentialAsync(
+        new DefaultAzureCredential());
+}
```
🤮

eerhardt replied:
LOL, I'm reverting this change anyway.

eerhardt committed:
- Fix an issue with the AzureResourcePreparer, it should copy default role assignments to applied role assignments even when empty. Add test to ensure.
- Update playground manifests
- Fix tests

eerhardt committed:
Add more tests

eerhardt committed:
Split all role assignments into separate bicep modules.

This works around an issue with CosmosDB where creating a managed identity and assigning a CosmosDB role assignment in the same bicep module fails.

captainsafia reviewed (approved):
This looks good given the gnarliness you had to work around with the way role assignments are managed for these resource types.

eerhardt committed:
Refactor how role assignments are created

Add a new virtual method on AzureProvisioningResource to add role assignments. This method is overriden in database resources (like CosmosDB and Redis) to customize how the role assignments are created.

eerhardt committed:
Clean up

eerhardt committed:
Fix tests

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into AzureDatabaseRoleAssignments2

eerhardt commented on src/Aspire.Hosting.Azure.AppContainers/AzureContainerAppsInfrastructure.cs:
```diff
@@ -314,96 +316,110 @@ private void ProcessRoleAssignments()
+                        customization(infra, azureResource);
+                    }
+
+                    // hook up any known parameters
```
I was able to refactor this so we don't need this hookup anymore. Instead, when the property is referenced, the parameter will get allocated.

eerhardt committed:
PR feedback

fill out XML docs

eerhardt commented on tests/Aspire.Hosting.Azure.Tests/AzureRedisExtensionsTests.cs:
```diff
@@ -1,23 +1,46 @@
+    [Theory]
+    [InlineData(false)]
+    [InlineData(true)]
+    public async Task AddAzureRedis(bool useAcaInfrastructure)
```
Added.

eerhardt committed:
PR feedback

captainsafia commented on tests/Aspire.Hosting.Azure.Tests/AzureFunctionsTests.cs:
```diff
@@ -271,17 +271,18 @@ public async Task AddAzureFunctionsProject_WorksWithAddAzureContainerAppsInfrast
 
         await ExecuteBeforeStartHooksAsync(app, default);
 
-        var projRoles = Assert.Single(model.Resources.OfType<AzureProvisioningResource>().Where(r => r.Name == $"funcapp-roles"));
```
Should we validate that the funcapp-identity resource emits the right manifest/bicep here? Particularly around the principal ID being an output that's consumed as a parameter elsewhere?

eerhardt replied:
I don't think that is necessary everywhere. We validate that in other places - specifically in the AzureContainerAppsTests.

https://github.com/dotnet/aspire/blob/122579052d497681ba156bb23058a70434eb1dd7/tests/Aspire.Hosting.Azure.Tests/AzureContainerAppsTests.cs#L2427-L2464

Feels like overkill to spread that all over.

Note that this test does verify that the principalId comes from the identity resource.

https://github.com/dotnet/aspire/blob/122579052d497681ba156bb23058a70434eb1dd7/tests/Aspire.Hosting.Azure.Tests/AzureFunctionsTests.cs#L286

captainsafia commented on src/Aspire.Hosting.Azure.CosmosDB/AzureCosmosDBExtensions.cs:
```diff
@@ -36,7 +36,8 @@ public static IResourceBuilder<AzureCosmosDBResource> AddAzureCosmosDB(this IDis
         var resource = new AzureCosmosDBResource(name, ConfigureCosmosDBInfrastructure);
-        return builder.AddResource(resource);
+        return builder.AddResource(resource)
+            .WithAnnotation(new DefaultRoleAssignmentsAnnotation(new HashSet<RoleDefinition>()));
```
What does an empty `DefaultRoleAssignmentsAnnotation` actually help us achieve here? This is just here to pass through to `AppliedRoleAssignmentsAnnotation` in the non-ACA scenario, right?

captainsafia replied:
> In ACA scenarios, the AzureResourcePreparer will copy the DefaultRoleAssignmentsAnnotation to a regular (empty) RoleAssignmentAnnotation, which will cause the role to be written to the role bicep file.

Ah, I [see where this happens](https://source.dot.net/#Aspire.Hosting.Azure/AzureResourcePreparer.cs,153).

> Thoughts?

IMO, an empty one feels better than a fake role assignment. We clarified in an offline discussion why we need `DefaultRoleAssignments` vs `RoleAssignment` here.

captainsafia reviewed (approved):
LGTM! Reads much easier than the annotations-based implementation. 😅

davidfowl commented on src/Aspire.Hosting.Azure.AppContainers/AzureContainerAppsInfrastructure.cs:
```diff
@@ -322,101 +321,69 @@ private void ProcessRoleAssignments()
+                    var context = new AddRoleAssignmentsContext(
+                        infra,
+                        roles,
+                        new(() => RoleManagementPrincipalType.ServicePrincipal),
+                        new(() => new BicepOutputReference("principalId", containerAppIdentityResource).AsProvisioningParameter(infra, parameterName: AzureBicepResource.KnownParameters.PrincipalId)),
+                        new(() => new BicepOutputReference("principalName", containerAppIdentityResource).AsProvisioningParameter(infra, parameterName: AzureBicepResource.KnownParameters.PrincipalName)));
```
Using Lazy in the public API is hella ugly. Would be cleaner if there was an `IRoleAssignmentsContext` and AddRoleAssignmentsContext as internal.

eerhardt replied:
Fixed in [Follow up on database role assignments (dotnet/aspire#8212)](https://github.com/dotnet/aspire/pull/8212)

eerhardt commented:
Merging now because this change and https://github.com/dotnet/aspire/pull/8171 will conflict. So merging will help avoid conflicts now.

I will address the above feedback in a follow up PR today.

eerhardt commented on src/Aspire.Hosting.Azure.AppContainers/AzureContainerAppsInfrastructure.cs:
```diff
@@ -322,101 +321,69 @@ private void ProcessRoleAssignments()
-                var existingResourceRoleAssignments = new List<AzureBicepResource>();
-                foreach (var (a, roles) in GetExistingResourceRoles())
+                var roleAssignments = new List<AzureBicepResource>();
+                foreach (var a in RoleAssignments.Keys)
```
Fixed in [Follow up on database role assignments (dotnet/aspire#8212)](https://github.com/dotnet/aspire/pull/8212).

eerhardt commented on src/Aspire.Hosting.Azure/AddRoleAssignmentsContext.cs:
```diff
@@ -0,0 +1,65 @@
+    public BicepValue<Guid> PrincipalId => _getPrincipalId.Value;
+
+    /// <summary>
+    /// Gets the principal name to use in the role assignment.
```
Added in [Follow up on database role assignments (dotnet/aspire#8212)](https://github.com/dotnet/aspire/pull/8212).
